### PR TITLE
Adjust ACO Tree to CakePHP 3.0

### PR DESCRIPTION
**Pull request by golgher**
Correcting the generation of ACOs tree for cakephp 3.0 (#48 and #32)

acl_extras [aco_update and aco_sync] will create nodes based on this: 
:rootNode/:plugin/:prefix/:controller/:action

### CODE REVIEW

**Comment by rchavik** [comment]
Aside from few minor CS issues, I think we need to see if `_getPrefixList()` can be done by the framework, ala `Router::prefixes()` in 2.x.

**Comment by markstory** [comment]
You might want to try iterating the connected routes instead of looking at the controllers and their namespaces.

**Comment by golgher** [comment]
I've updated AclExtras to iterate connected routes and use the route template to create ACOs.

File: src/AclExtras.php [hunk 1]
```diff
@@ -138,25 +162,117 @@ public function acoUpdate($params = [])
...
         if (empty($params['plugin'])) {
-            $controllers = $this->getControllerList();
-            $this->_updateControllers($root, $controllers);
+            $this->_processControllers($root);
+            $this->_processPrefixes($root);
             $plugins = $this->_getPluginList();
         } else {
             $plugin = $params['plugin'];
-            if (!Plugin::loaded($plugin)) {
+            if (!in_array($plugin, App::objects('plugin')) || !CakePlugin::loaded($plugin)) {
```
Comment (markstory) [inline src/AclExtras.php]: App::objects() does not exist in CakePHP 3.0. Nor does CakePlugin.

**Comment by golgher** [comment]
I've updated outdated code, but I found 2 problems when plugin passed as arg:
1. In cakephp/src/Console/Shell.php lines 356 to 358:
```php
    if (!empty($this->params['plugin'])) {
        Plugin::load($this->params['plugin']);
    }
```
When a plugin with routes is used, this routes isn't loaded. I tried to comment these lines, and everything work well. 

2. When app have a prefix route been used for app and plugin, like this example:
```php
Router::prefix('admin', function ($routes) {
    $routes->plugin('Test', function ($routes) {
        $routes->connect('/:controller');
        $routes->fallbacks('InflectedRoute');
    });
    $routes->fallbacks('InflectedRoute');
});
```
When I tried to synchronise ACOs, this version deleted app's ACOS for this prefix rule, updating plugin correctly.

**Comment by boindil** [comment]
Any progress here? CakePHP 3.0 is out and this seems to be an issue still?

Currently the generated DB-entries would still be wrong when doing aco sync having stuff like this ```:rootNode/:plugin/:prefix/:controller/:action```, would they?

File: src/AclExtras.php [hunk 2]
```diff
@@ -148,15 +173,108 @@ public function acoUpdate($params = [])
                 return false;
             }
             $plugins = [$params['plugin']];
+            $this->_processPlugins($root,$plugins);
+            $this->foundACOs = array_slice($this->foundACOs, 1, NULL, true);
```
Comment (markstory) [inline src/AclExtras.php]: NULL should be null.

File: src/AclExtras.php [hunk 3]
```diff
@@ -148,15 +173,108 @@ public function acoUpdate($params = [])
...
             $controllers = $this->getControllerList($plugin);
             $path = $this->rootNode . '/' . $plugin;
-            $pluginRoot = $this->_checkNode($path, preg_replace('/\//', '\\', Inflector::camelize($plugin)), $root->id);
-            $this->_updateControllers($pluginRoot, $controllers, $plugin);
+            $pathNode = $this->_checkNode($path, $plugin, $root->id);
+            $this->foundACOs[$root->id][] = $plugin;
+            $template = '/' . $plugin . '/:controller';
+            if(isset($this->foundACOs[$pathNode->id])){
+                $this->foundACOs[$pathNode->id] += $this->_updateControllers($pathNode, $controllers, $template, $plugin);
+            } else{
```
Comment (markstory) [inline src/AclExtras.php]: `if (` and `else {`.

**Comment by markstory** [comment]
@boindil You right, this pull request hasn't been merged. There are a bunch of coding standards related issues and it would be great to have some additional tests to ensure we don't have future regressions. If you need these fixes perhaps you could help out?

**Comment by boindil** [comment]
@markstory I don't know if I'm the right one to ask for doing this, but I'll have a look and try. :-D

**Comment by dakota** [comment]
Closing in favour of #68

**Comment by dakota** [comment]
Thanks @golgher for the fix!

**Pull request closed without merging**